crates/invoker-impl/src/invocation_task.rs

```diff
@@ -409,46 +405,146 @@ where
                     .deployment_metadata_resolver
-                    .get_deployment(&deployment_id)
-                    .ok_or_else(|| InvocationTaskError::UnknownDeployment(deployment_id)));
-                (deployment_metadata, /* has_changed= */ false)
+                    .get_deployment(&chosen_deployment.deployment_id)
+                    .ok_or_else(|| InvocationTaskError::UnknownDeployment(
+                        chosen_deployment.deployment_id
+                    )));
+
+                // todo: We should support resuming an invocation with a newer protocol version if
+                //  the endpoint supports it
+                if !ServiceProtocolVersion::is_supported(chosen_deployment.service_protocol_version)
+                {
+                    shortcircuit!(Err(InvocationTaskError::UnsupportedServiceProtocolVersion(
+                        chosen_deployment.service_protocol_version
+                    )));
+                }
+
+                (
+                    deployment_metadata,
+                    chosen_deployment.service_protocol_version,
+                    /* has_changed= */ false,
+                )
             } else {
                 // We can choose the freshest deployment for the latest revision
                 // of the registered service.
                 let deployment = shortcircuit!(self
                     .deployment_metadata_resolver
                     .resolve_latest_deployment_for_service(self.invocation_target.service_name())
                     .ok_or(InvocationTaskError::NoDeploymentForService));
-                (deployment, /* has_changed= */ true)
+
+                let chosen_service_protocol_version =
+                    shortcircuit!(ServiceProtocolVersion::choose_max_supported_version(
+                        &deployment.metadata.supported_protocol_versions,
+                    )
+                    .ok_or_else(|| {
+                        InvocationTaskError::IncompatibleServiceEndpoint(
+                            deployment.id,
+                            deployment.metadata.supported_protocol_versions.clone(),
+                        )
+                    }));
+
+                (
+                    deployment,
+                    chosen_service_protocol_version,
+                    /* has_changed= */ true,
+                )
             };
 
-        self.send_invoker_tx(InvocationTaskOutputInner::SelectedDeployment(
-            deployment.id,
+        self.send_invoker_tx(InvocationTaskOutputInner::ChosenDeployment(
+            ChosenDeployment::new(deployment.id, chosen_service_protocol_version),
             deployment_changed,
         ));
 
+        // create a versioned connection
+        let invocation_connection =
+            VersionedInvocationConnection::new(self, chosen_service_protocol_version);
+
+        invocation_connection
+            .run(journal_metadata, deployment, journal_stream, state_iter)
+            .await
+    }
```

Comment:
Maybe an idea to improve the code org here. I can see two distinct phases of this invocation task: prepare and run. Prepare loads some stuff from rocksdb and schema registry, run actually executes the invocation. Perhaps one way to make this more evident is to rename `InvocationTask#run_internal` in `prepare_and_run`.

Reply:
TBF even splitting to two struct seems not needed now, you could just have a function `run_v1` that takes the encoder/decoder as input args. If something changes in the next iterations most likely it won't require rewriting that struct, so perhaps let's split only if we need to do it?

Reply:
If I did it this way, then I would have to pass the encoder/decoder and the chosen service protocol through all methods to the places where they are needed. I think it makes sense to instantiate a versioned connection/whatever which is bound to a specific service protocol version and has the encoder/decoder properly configured.

Reply:
> Maybe an idea to improve the code org here. I can see two distinct phases of this invocation task: prepare and run. Prepare loads some stuff from rocksdb and schema registry, run actually executes the invocation. Perhaps one way to make this more evident is to rename InvocationTask#run_internal in prepare_and_run.

I am not sure whether it is important to expose how `run_internal` is internally implemented. From a users perspective nothing has changed compared to before just that internally we first figure out the service protocol and then use a helper struct to create the service protocol specific components.

Reply:
Nothing changes from user perspective agree. My comment was mostly about how we organize this code, which is quite a crucial piece of the system :)

Reply:
I'll check whether I can make it clearer.

Reply:
I've renamed it into `select_protocol_version_and_run`.